# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/vm/Shape.h
# Commit: bad57461dcc4
# Full Hash: bad57461dcc4c6c52064bf985ee737d8abbc183b
# Author: Jan de Mooij <jdemooij@mozilla.com>
# Date: 2021-03-23 21:43:59
# Regressor Bug: 1696178
# File Overlap Count: 1
# Description:
#   Bug 1696178 part 5 - Add ObjectFlag::HasNonWritableOrAccessorExcludingProto. r=jonco
#   
#   This will be used by the next patch to optimize Object.assign.
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D108407
# ==============================================================================

diff -r 12574c8b209d -r bad57461dcc4 js/src/vm/Shape.h
--- a/js/src/vm/Shape.h	Tue Mar 23 12:43:56 2021 +0000
+++ b/js/src/vm/Shape.h	Tue Mar 23 12:43:57 2021 +0000
@@ -657,6 +657,19 @@
 
   // See JSObject::isQualifiedVarObj().
   QualifiedVarObj = 1 << 8,
+
+  // If set, the object may have a non-writable property or an accessor
+  // property.
+  //
+  // * This is only set for PlainObjects because we only need it for these
+  //   objects and setting it for other objects confuses insertInitialShape.
+  //
+  // * This flag does not account for properties named "__proto__". This is
+  //   because |Object.prototype| has a "__proto__" accessor property and we
+  //   don't want to include it because it would result in the flag being set on
+  //   most proto chains. Code using this flag must check for "__proto__"
+  //   property names separately.
+  HasNonWritableOrAccessorPropExclProto = 1 << 9,
 };
 
 using ObjectFlags = EnumFlags<ObjectFlag>;
